x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.authc.saml;
+
+import org.elasticsearch.common.unit.TimeValue;
+import org.opensaml.saml.saml2.core.LogoutResponse;
+import org.w3c.dom.Element;
+
+import java.time.Clock;
+import java.util.Collection;
+
+import static org.elasticsearch.xpack.security.authc.saml.SamlUtils.samlException;
+
+public class SamlLogoutResponseHandler extends SamlResponseHandler {
+
+    private static final String LOGOUT_RESPONSE_TAG_NAME = "LogoutResponse";
+
+    public SamlLogoutResponseHandler(
+        Clock clock, IdpConfiguration idp, SpConfiguration sp, TimeValue maxSkew) {
+        super(clock, idp, sp, maxSkew);
+    }
+
+    public void handle(byte[] payload, Collection<String> allowedSamlRequestIds) {
+        final Element root = parseSamlMessage(payload);
+        if (LOGOUT_RESPONSE_TAG_NAME.equals(root.getLocalName()) && SAML_NAMESPACE.equals(root.getNamespaceURI())) {
+            final LogoutResponse logoutResponse = buildXmlObject(root, LogoutResponse.class);
+            if (logoutResponse == null) {
+                throw samlException("Cannot convert element {} into LogoutResponse object", root);
+            }
+            if (logoutResponse.isSigned()) {
```

Comment:
The response _MUST_ be signed ([4.4.4.2](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf))  so we should fail if it is not signed.

Reply:
The spec says "Must authenticate itself either by signing or a binding specific mechansim". While the HTTP-POST `LogoutResponse` is signed, the `LogoutResponse` from HTTP-Redirect binding is _not_. So I would prefer it to be optional.

Reply:
> or a binding specific mechanism

this alludes to the artifact binding that is not used/supported here

> . While the HTTP-POST LogoutResponse is signed, the LogoutResponse from HTTP-Redirect binding is not.

Why are you stating that as a fact ? It can very well be signed and according to the specification it `MUST` be.

Reply:
> Why are you stating that as a fact ?
>
It's not an universal fact. I was trying to avoid explicitly mentioning a popular idP's name. Sorry about the confusion. If we mandate signature verification and Kibana starts to redirect the response to ES, the verification will fail since it does not have signature.

In addition, including signature in HTTP-Redirect request may risk hitting the (practical) limit of URL length.

Reply:
>  If we mandate signature verification and Kibana starts to redirect the response to ES, the verification will fail since it does not have signature.

It all boils down to how permissive we want to be with things that go against the specification. If a popular IDP doesn't adhere to the specification, do we happily consume what they are sending us or do we not ? And if another popular IDP comes along and doesn't adhere to the standards in a different way, do we happily accept that too? Where do we draw the line? I believe that drawing the line at what the specification says is easier to argue about than drawing the line in an arbitrary leniency we deem acceptable at some point. 

> In addition, including signature in HTTP-Redirect request may risk hitting the (practical) limit of URL length. I guess this is probably the reason why it is not signed by the aforementioned idP

I doubt that. SAML Logout Response messages are quite small ( comparable to authentication requests that also use the HTTP Redirect binding and are signed )

Reply:
> I believe that drawing the line at what the specification says is easier to argue about than drawing the line in an arbitrary leniency we deem acceptable at some point.
> 
This I agree. But the whole story is not only just with a particular idP. The leniency was already implemented by us (by dropping the `LogoutResponse` at Kibana side). We could choose to tighten up the behaviour and accept the fact it might break user's existing setup. I am happy to conform the standard if we are certain that is the price we would like to pay.

> authentication requests that also use the HTTP Redirect binding and are signed
> 
The auth `Response` from idP is signed but uses HTTP-POST (in fact we only support HTTP-POST binding for this). Our SAML AuthRequest uses redirect and it is not signed. As far as I know, Apache server has 4096 (or is it 8192) limit on URL length, it is a thing when URL gets too long.

I have to admit that I only just started to read the spec and have many knownlege gaps. For an example, the signature part is confusing to me. It mandates a `LogoutResponse` to be signed, but signature for authn `Response` is optional. Isn't the authn `Response` a more sensitive/valuable information compared to `LogoutResponse`?

Reply:
> The leniency was already implemented by us (by dropping the LogoutResponse at Kibana side).

This was not leniency. This was part of the specification not handled. If we _do_ handle it, let's do it correctly.

> The auth Response from idP is signed but uses HTTP-POST (in fact we only support HTTP-POST binding for this)

True, but I said 

> authentication requests that also use the HTTP Redirect binding and are signed

and SAML Authentication Requests are _not_ SAML Authentication Responses :)

>  Our SAML AuthRequest uses redirect and it is not signed. 

It is when it is configured to be https://www.elastic.co/guide/en/elasticsearch/reference/master/saml-guide-authentication.html#_configuring_elasticsearch_for_signing

> As far as I know, Apache server has 4096 (or is it 8192) limit on URL length, it is a thing when URL gets too long.

And SAML with authentication requests ( using the HTTP Redirect binding) being signed has been out there for 15 years now, this is heavily supported and used, I don't think we need to worry about this. 

> but signature for authn Response is optional

I don't think this is true.

Reply:
First of all, thanks for indulging me with the discussions. I really appreciate it. It helps me understanding the issue better ❤️ 
> It is when it is configured to be
>
Thanks for pointing this out. I completely ignored this part when setting up saml locally. But isn't it actually an example that we do allow unsigned `LogoutResponse` (the difference being we are the sender in this case)? 

> SAML with authentication requests ( using the HTTP Redirect binding) being signed has been out there for 15 years now, this is heavily supported and used
>
This is a good point as well. Given authn request can be signed and use HTTP Redirect, I agree there will be no issue for `LogoutResponse`.

> I don't think this is true.
>
You are right that it is not "optional". The word used in the spec [6.4.2](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf) is "SHOULD". It is stronger than optional, but not MUST. Our implementation for this part treats it as optional:
https://github.com/elastic/elasticsearch/blob/a443e1387dc262b3ec58b53c6b2f2001a781093a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlAuthenticator.java#L90-L92

Reply:
> But isn't it actually an example that we do allow unsigned LogoutResponse (the difference being we are the sender in this case)?

I don't remember the discussions we had while implementing this , I can assume we made a decision that interoperability is more important than specification adherence at that time. We can revisit that decision too :) 


> You are right that it is not "optional". The word used in the spec 6.4.2 is "SHOULD".

We only support the web browser single sign on profile, so 4.1 in that spec, see 4.1.3.5


>  Our implementation for this part treats it as optional:

It -hopefully- doesn't!!! A SAML response contains SAML assertions. What we do support is either the response or the assertion is signed :) There must be an integrity protection implemented, otherwise it would be trivial for attackers to bypass authentication

Reply:
> We only support the web browser single sign on profile, so 4.1 in that spec, see 4.1.3.5
> 
You are right. I read the wrong section 🤦 

> What we do support is either the response or the assertion is signed
> 
Yes that's what I figured as well. There is no risk in this part. 

I think the only thing that bothers me is this "popular idP". The original issue #43264 is created with it as an example of why we consider to support HTTP-POST LogoutResponse. So it feels rather unsatifying if we fix one binding and breaks the other (the current working one).

The code change itself either way is trivial. The decision making part is more important. I feel we could talk about this over sync time.

Reply:
> I think the only thing that bothers me is this "popular idP".

I don't think we need to make changes that satisfy any "popular IDP", we should make changes that make sense for the users and are correct. What's more, I see no indication whatsoever that any IDP doesn't support signed Logout Responses so I'm not sure what we would be actually breaking with this. 

I think we should not be lenient here for no reason, but happy to discuss this more on the sync.

Reply:
Sync update: we decided to conform to the spec and perform full validation. Caller of the API can decide what to do with validation errors if any. Thanks for the discussion.